app/upgrades/v20/upgrades.go

```diff
@@ -32,6 +45,122 @@ func CreateUpgradeHandler(
+		if err != nil {
+			return err
+		}
+
+		// At the time of v20 upgrade, we only have concentrated pools
+		// that are linked to balancer. Concentrated gauges receive all rewards
+		// and then retroactively update balancer.
+		// Concentrated pools have NoLock Gauge associated with them.
+		// As a result, we look for this specific type here.
+		// If type mismatched, this is a CFMM pool gauge. In that case,
+		// we continue to the next incentive record after validating
+		// that there is no migration record present for this CFMM pool. That is
+		// it is not incentivized individually when concentrated pool already retroactively
+		// distributed rewards to it.
+		if gauge.DistributeTo.LockQueryType != lockuptypes.NoLock {
+			// Validate that if there is a migration record pair between a concentrated
+			// and a cfmm pool, only concentrated is present in the distribution records.
+			longestLockableDuration, err := keepers.PoolIncentivesKeeper.GetLongestLockableDuration(ctx)
+			if err != nil {
+				return err
+			}
+			cfmmPoolID, err := keepers.PoolIncentivesKeeper.GetPoolIdFromGaugeId(ctx, gaugeID, longestLockableDuration)
+			if err != nil {
+				return err
+			}
+
+			// If we had a migration record present and a balancer pool is still incentivized individually,
+			// something went wrong. This is because the presence of migration record implies retroactive
+			// incentive distribution from concentrated to balancer.
+			linkedConcentratedPoolID, hasAssociatedConcentratedPoolLinked := poolIDMigrationRecordMap[cfmmPoolID]
+			if hasAssociatedConcentratedPoolLinked {
+				return IncentivizedCFMMDirectWhenMigrationLinkPresentError{
+					CFMMPoolID:         cfmmPoolID,
+					ConcentratedPoolID: linkedConcentratedPoolID,
+					CFMMGaugeID:        gaugeID,
+				}
+			}
+
+			// Validation passed. This was an individual CFMM pool with no link to concentrated
+			// Silently skip it.
+			continue
+		}
+
+		// Get PoolID associated with the given NoLock gauge ID
+		// NoLock gauges are associated with an incentives epoch duration.
+		incentivesEpochDuration := keepers.IncentivesKeeper.GetEpochInfo(ctx).Duration
+		concentratedPoolID, err := keepers.PoolIncentivesKeeper.GetPoolIdFromGaugeId(ctx, gaugeID, incentivesEpochDuration)
+		if err != nil {
+			return err
+		}
+
+		associatedGammPoolID, ok := poolIDMigrationRecordMap[concentratedPoolID]
+		if !ok {
+			// There is no CFMM pool ID for the concentrated pool ID, contniue to the next.
+			continue
+		}
+
+		// Found concentrated and CFMM pools that are linked by
+		// migration records. Create a Group for them
+		groupedPoolIDs := []uint64{concentratedPoolID, associatedGammPoolID}
+		groupGaugeID, err := keepers.IncentivesKeeper.CreateGroupAsIncentivesModuleAcc(ctx, incentivestypes.PerpetualNumEpochsPaidOver, groupedPoolIDs)
+		if err != nil {
+			return err
+		}
+
+		// Replace the gauge ID with the group gauge ID in the distribution records
+		distrInfo.Records[i].GaugeId = groupGaugeID
```

Comment:
So i guess we will continue to label these durations for groups as epochDuration rather than zero right?

Reply:
I'm not sure how zero epochDurations relate to groups. Could you elaborate?

We previously discussed zero durations in the context of NoLock gauges. If that's what you're referring to, we will not refactor pre-existing logic at this time since that is a non-goal volume splitting